ddtrace/tracer/rules_sampler.go

```diff
@@ -0,0 +1,572 @@
+// effective rate at the time it is called. The effective rate is computed by averaging the rate
+// for the previous second with the current rate
+func (r *rateLimiter) allowOne(now time.Time) (bool, float64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if d := now.Sub(r.prevTime); d >= time.Second {
+		// enough time has passed to reset the counters
+		if d.Truncate(time.Second) == time.Second && r.seen > 0 {
+			// exactly one second, so update prev
+			r.prevAllowed = r.allowed
+			r.prevSeen = r.seen
+		} else {
+			// more than one second, so reset previous rate
+			r.prevAllowed = 0
+			r.prevSeen = 0
+		}
+		r.prevTime = now
+		r.allowed = 0
+		r.seen = 0
+	}
+
+	r.seen++
+	var sampled bool
+	if r.limiter.AllowN(now, 1) {
+		r.allowed++
+		sampled = true
+	}
+	er := (r.prevAllowed + r.allowed) / (r.prevSeen + r.seen)
+	return sampled, er
+}
+
+// newSingleSpanRateLimiter returns a rate limiter which restricts the number of single spans sampled per second.
+// This defaults to infinite, allow all behaviour. The MaxPerSecond value of the rule may override the default.
+func newSingleSpanRateLimiter(mps float64) *rateLimiter {
+	limit := math.MaxFloat64
+	if mps > 0 {
+		limit = mps
+	}
+	return &rateLimiter{
+		limiter:  rate.NewLimiter(rate.Limit(limit), int(math.Ceil(limit))),
+		prevTime: time.Now(),
+	}
+}
+
+// globMatch compiles pattern string into glob format, i.e. regular expressions with only '?'
+// and '*' treated as regex metacharacters.
+func globMatch(pattern string) (*regexp.Regexp, error) {
+	// escaping regex characters
+	pattern = regexp.QuoteMeta(pattern)
+	// replacing '?' and '*' with regex characters
+	pattern = strings.Replace(pattern, "\\?", ".", -1)
+	pattern = strings.Replace(pattern, "\\*", ".*", -1)
+	//pattern must match an entire string
+	return regexp.Compile(fmt.Sprintf("^%s$", pattern))
+}
+
+// samplingRulesFromEnv parses sampling rules from the DD_TRACE_SAMPLING_RULES,
+// DD_SPAN_SAMPLING_RULES and DD_SPAN_SAMPLING_RULES_FILE environment variables.
+func samplingRulesFromEnv() ([]SamplingRule, error) {
+	var errs []string
+	var rules []SamplingRule
+	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
+	if rulesFromEnv != "" {
+		traceRules, err := unmarshalSamplingRules([]byte(rulesFromEnv), SamplingRuleTrace)
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+		rules = append(rules, traceRules...)
+	}
+	spanRules, err := unmarshalSamplingRules([]byte(os.Getenv("DD_SPAN_SAMPLING_RULES")), SamplingRuleSpan)
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+	rules = append(rules, spanRules...)
+	rulesFile := os.Getenv("DD_SPAN_SAMPLING_RULES_FILE")
+	if len(rules) != 0 {
+		if rulesFile != "" {
+			log.Warn("DIAGNOSTICS Error(s): DD_SPAN_SAMPLING_RULES is available and will take precedence over DD_SPAN_SAMPLING_RULES_FILE")
+		}
+		if len(errs) != 0 {
+			return rules, fmt.Errorf("%s", strings.Join(errs, "\n\t"))
```

Comment:
Rather than duplicating this `strings.Join` a bunch of times throughout the code (which will quickly become brittle if we decide to change how they are joined), can we try to do that one time after the function is done? 
I wonder if something like this could work: https://go.dev/play/p/5CRPMhI9hA4

There's probably a less hacky way to do this, using ifs or helper functions, but just wanted to throw out an idea.